llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp

```diff
@@ -82,6 +83,25 @@ static bool ShouldSignWithBKey(const Function &F, const AArch64Subtarget &STI) {
   return Key == "b_key";
 }
 
+static bool hasELFSignedGOTHelper(const Function &F,
+                                  const AArch64Subtarget *STI) {
+  if (!Triple(STI->getTargetTriple()).isOSBinFormatELF())
+    return false;
+  const Module *M = F.getParent();
+  uint64_t PAuthABIPlatform = -1;
+  if (const auto *PAP = mdconst::extract_or_null<ConstantInt>(
+          M->getModuleFlag("aarch64-elf-pauthabi-platform")))
+    PAuthABIPlatform = PAP->getZExtValue();
+  if (PAuthABIPlatform != ELF::AARCH64_PAUTH_PLATFORM_LLVM_LINUX)
+    return false;
+  uint64_t PAuthABIVersion = -1;
+  if (const auto *PAV = mdconst::extract_or_null<ConstantInt>(
+          M->getModuleFlag("aarch64-elf-pauthabi-version")))
+    PAuthABIVersion = PAV->getZExtValue();
+  return (PAuthABIVersion &
+          (1 << ELF::AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_GOT)) != 0;
```

Comment:
The function seems to return `true` if there is no `aarch64-elf-pauthabi-version` flag, which doesn't seem right?

```suggestion
  const auto *PAV = mdconst::extract_or_null<ConstantInt>(
          M->getModuleFlag("aarch64-elf-pauthabi-version"));
  if (!PAV)
      return false;
  return PAV->getZExtValue() &
          (1 << ELF::AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_GOT);
```

Reply:
You are right, that's not correct, thanks for pointing this out. This magic with -1 was OK in AsmPrinter (probably should be changes to `std::optional` or smth there as well though), but here it's not applicable. Applied your suggestion in 4c85d6c31cfc89e8543d4f8d78c9cd82dd33ca37